File: src/main/resources/assets/minecraft/scripts/RenderART500M.js
var renderClass = "jp.ngt.rtm.render.VehiclePartsRenderer";
importPackage(Packages.org.lwjgl.opengl);
importPackage(Packages.jp.ngt.rtm.render);
importPackage(Packages.jp.ngt.ngtlib.math);
importPackage(Packages.com.github.ikoko125.banetsu2419_ns3000.model);

function init(modelSet, modelObject) {
    ModelART500M.registerObjects(renderer);
}





function randomcurtain(entity){
	var time = renderer.getMCTime();
	if(time < 400 || time > 13000){
		nightWindow.render(renderer);
	}
}


function render_door(entity){
	var moveL = 0;
	var moveR = 0;
	var doorMoveL = renderer.getDoorMovementL(entity),
		doorMoveR = renderer.getDoorMovementR(entity);
		if(entity != null){
			moveL = renderer.sigmoid(entity.doorMoveL / 5);
			moveR = renderer.sigmoid(entity.doorMoveR / 5);
			}

			if(time < 400 || time > 13000){
		  GL11.glPushMatrix();
		   GL11.glTranslatef(moveL * 0.07, 0.0, doorMoveL * 0.7);
		   doorLF1.render(renderer);
		  GL11.glPopMatrix();
		  GL11.glPushMatrix();
		   GL11.glTranslatef(moveL * 0.07, 0.0, -doorMoveL * 0.7);
		   doorLB1.render(renderer);
		  GL11.glPopMatrix();
		  GL11.glPushMatrix();
		   GL11.glTranslatef(-moveR * 0.07, 0.0, doorMoveR * 0.7);
		   doorRF1.render(renderer);
		  GL11.glPopMatrix();
		  GL11.glPushMatrix();
		   GL11.glTranslatef(-moveR * 0.07, 0.0, -doorMoveR * 0.7);
		   doorRB1.render(renderer);
		  GL11.glPopMatrix();
		}
		else{
		  GL11.glPushMatrix();
		   GL11.glTranslatef(moveL * 0.07, 0.0, doorMoveL * 0.7);
		   doorLF.render(renderer);
		  GL11.glPopMatrix();
		  GL11.glPushMatrix();
		   GL11.glTranslatef(moveL * 0.07, 0.0, -doorMoveL * 0.7);
		   doorLB.render(renderer);
		  GL11.glPopMatrix();
		  GL11.glPushMatrix();
		   GL11.glTranslatef(-moveR * 0.07, 0.0, doorMoveR * 0.7);
		   doorRF.render(renderer);
		  GL11.glPopMatrix();
		  GL11.glPushMatrix();
		   GL11.glTranslatef(-moveR * 0.07, 0.0, -doorMoveR * 0.7);
		   doorRB.render(renderer);
		  GL11.glPopMatrix();
		}
	
}	





	
function render(entity, pass, par3){
	time = renderer.getMCTime();
	if(pass==0){
		body.render(renderer);
		render_door(entity);
	}
	if(pass==1){
		body.render(renderer);
		randomcurtain(entity);
		render_door(entity);
	}
	if(pass==2){
		body.render(renderer);
		randomcurtain(entity);
		render_door(entity);
	}
	if(pass==3){
		body.render(renderer);
		randomcurtain(entity);
		render_door(entity);
	}
	if(pass==4){
		body.render(renderer);
		randomcurtain(entity);
		render_door(entity);
	}


		//パンタ動作用
        if (pass >= 0 && pass <= 4) {
			if(entity == null){
				return;
			}
			else{
				var formation = entity.getFormation();
			if(formation == null) return;
			var CarNumber = formation.getEntry(entity).entryId+1;
	
			switch( CarNumber ) {
			case 1:
				GL11.glPushMatrix();
				car1.render(renderer);
				GL11.glPopMatrix();
				break;
			case 2:
				GL11.glPushMatrix();
				car2.render(renderer);
				GL11.glPopMatrix();
				break;
			case 3:
				GL11.glPushMatrix();
				car3.render(renderer);
				GL11.glPopMatrix();
				break;
			case 4:
				GL11.glPushMatrix();
				car4.render(renderer);
				GL11.glPopMatrix();
				break;
			case 5:
				GL11.glPushMatrix();
				car5.render(renderer);
				GL11.glPopMatrix();
				break;
			case 6:
				GL11.glPushMatrix();
				car6.render(renderer);
				GL11.glPopMatrix();
				break;
			case 7:
				GL11.glPushMatrix();
				car7.render(renderer);
				GL11.glPopMatrix();
				break;
			case 8:
				GL11.glPushMatrix();
				car8.render(renderer);
				GL11.glPopMatrix();
				break;
			case 9:
				GL11.glPushMatrix();
				car9.render(renderer);
				GL11.glPopMatrix();
				break;
			case 10:
				GL11.glPushMatrix();
				car10.render(renderer);
				GL11.glPopMatrix();
				break;
			}
			}
		}

		if (pass >= 0 && pass <= 4) {
			var formation = entity.getFormation();
		if(formation == null) return;
		var CarNumber = formation.getEntry(entity).entryId+1;

		switch( CarNumber ) {
		case 1:
			GL11.glPushMatrix();
			   car1.render(renderer);
			   GL11.glPopMatrix();
			break;
		case 2:
			GL11.glPushMatrix();
			   car2.render(renderer);
			   GL11.glPopMatrix();
			break;
		case 3:
			GL11.glPushMatrix();
			   car3.render(renderer);
			   GL11.glPopMatrix();
			break;
		case 4:
			GL11.glPushMatrix();
			   car4.render(renderer);
			   GL11.glPopMatrix();
			break;
		case 5:
			GL11.glPushMatrix();
			   car5.render(renderer);
			   GL11.glPopMatrix();
			break;
		case 6:
			GL11.glPushMatrix();
			   car6.render(renderer);
			   GL11.glPopMatrix();
			break;
		case 7:
			GL11.glPushMatrix();
			   car7.render(renderer);
			   GL11.glPopMatrix();
			break;
		case 8:
			GL11.glPushMatrix();
			   car8.render(renderer);
			   GL11.glPopMatrix();
			break;
		case 9:
			GL11.glPushMatrix();
			   car9.render(renderer);
			   GL11.glPopMatrix();
			break;
		case 10:
			GL11.glPushMatrix();
			   car10.render(renderer);
			   GL11.glPopMatrix();
			break;
		}
		}


		var pantaRotate1,pantaRotate2,pantaRotate4;//パンタ動作用
		try{
		  pantaRotate1 = renderer.sigmoid(entity.pantograph_F / 60) * 1.2 * 30;//パンタ動作用
		  pantaRotate2 = renderer.sigmoid(entity.pantograph_F / 60) * 1.2 * 56;//パンタ動作用
		  pantaRotate4 = renderer.sigmoid(entity.pantograph_F / 60) * 1.2 * 66;//パンタ動作用
		}catch(e)
		{}
		if(pass == 0){
			GL11.glPushMatrix();
			renderer.rotate(pantaRotate1, 'X', 0.0, 2.2085, 2.3091);
			panta1.render(renderer);
			renderer.rotate(-pantaRotate2, 'X', 0.0, 3.0606, 3.8144);
			panta2.render(renderer);
			renderer.rotate(pantaRotate1, 'X', -0.0320, 4.0304, 2.3787);
			panta3.render(renderer);
			GL11.glPopMatrix();
			GL11.glPushMatrix();
			renderer.rotate(pantaRotate1, 'X', 0.0, 2.2085, 2.3091);
			renderer.rotate(-pantaRotate2, 'X', 0.0, 3.0606, 3.8144);
			renderer.rotate(pantaRotate4, 'X', 0.0, 2.22, 3.0478);
			panta4.render(renderer);
			GL11.glPopMatrix();
		}
		
}

File: src/main/kotlin/com/github/ikoko125/banetsu2419_ns3000/render/RenderART500M.kt
package com.github.ikoko125.banetsu2419_ns3000.render

import com.github.ikoko125.banetsu2419_ns3000.model.ModelART500M
import jp.ngt.ngtlib.math.NGTMath
import jp.ngt.rtm.entity.train.EntityTrainBase
import jp.ngt.rtm.render.VehiclePartsRenderer
import org.lwjgl.opengl.GL11

@Suppress("unused")
object RenderART500M {
    @Suppress("DuplicatedCode")
    @JvmStatic
    fun render(renderer: VehiclePartsRenderer, entity: EntityTrainBase?, pass: Int) {
        val model = ModelART500M.getModel(renderer.modelName)!!

        model.body.render(renderer)
        renderDoor(renderer, entity)
        if (pass in 1..4) renderRandomCurtain(renderer)

        if (entity == null) return

        if (pass in 0..4) {
            GL11.glPushMatrix()

            when (entity.formation.getEntry(entity).entryId + 1) {
                1 -> model.car1.render(renderer)
                2 -> model.car2.render(renderer)
                3 -> model.car3.render(renderer)
                4 -> model.car4.render(renderer)
                5 -> model.car5.render(renderer)
                6 -> model.car6.render(renderer)
                7 -> model.car7.render(renderer)
                8 -> model.car8.render(renderer)
                9 -> model.car9.render(renderer)
                10 -> model.car10.render(renderer)
            }

            GL11.glPopMatrix()
        }

        if (pass != 0) return

        val pantoRotate1 = (NGTMath.sigmoid(entity.pantograph_F / 60.0, 5.0) * 1.2 * 30).toFloat()
        val pantoRotate2 = (NGTMath.sigmoid(entity.pantograph_F / 60.0, 5.0) * 1.2 * 56).toFloat()
        val pantoRotate4 = (NGTMath.sigmoid(entity.pantograph_F / 60.0, 5.0) * 1.2 * 66).toFloat()

        GL11.glPushMatrix()

        renderer.rotate(pantoRotate1, 'X', 0.0f, 2.2085f, 2.3091f)
        model.panta1.render(renderer)

        renderer.rotate(-pantoRotate2, 'X', 0.0f, 3.0606f, 3.8144f)
        model.panta2.render(renderer)

        renderer.rotate(pantoRotate1, 'X', -0.032f, 4.0304f, 2.3787f)
        model.panta3.render(renderer)

        GL11.glPopMatrix()

        GL11.glPushMatrix()
        renderer.rotate(pantoRotate1, 'X', 0.0f, 2.2085f, 2.3091f)
        renderer.rotate(-pantoRotate2, 'X', 0.0f, 3.0606f, 3.8144f)
        renderer.rotate(pantoRotate4, 'X', 0.0f, 2.22f, 3.0478f)
        model.panta4.render(renderer)
        GL11.glPopMatrix()
    }

    @Suppress("DuplicatedCode")
    private fun renderDoor(renderer: VehiclePartsRenderer, entity: EntityTrainBase?) {
        val moveL = if (entity != null) NGTMath.sigmoid(entity.doorMoveL / 5.0, 5.0).toFloat() else 0.0f
        val moveR = if (entity != null) NGTMath.sigmoid(entity.doorMoveR / 5.0, 5.0).toFloat() else 0.0f
        val doorMoveL = renderer.getDoorMovementL(entity)
        val doorMoveR = renderer.getDoorMovementR(entity)
        val time = renderer.mcTime
        val model = ModelART500M.getModel(renderer.modelName)!!

        if (time !in 401..12999) {
            GL11.glPushMatrix()
            GL11.glTranslatef(moveL * 0.07f, 0.0f, doorMoveL * 0.7f)
            model.doorFL1.render(renderer)
            GL11.glPopMatrix()

            GL11.glPushMatrix()
            GL11.glTranslatef(moveL * 0.07f, 0.0f, -doorMoveL * 0.7f)
            model.doorBL1.render(renderer)
            GL11.glPopMatrix()

            GL11.glPushMatrix()
            GL11.glTranslatef(-moveR * 0.07f, 0.0f, doorMoveR * 0.7f)
            model.doorFR1.render(renderer)
            GL11.glPopMatrix()

            GL11.glPushMatrix()
            GL11.glTranslatef(-moveR * 0.07f, 0.0f, -doorMoveR * 0.7f)
            model.doorBR1.render(renderer)
            GL11.glPopMatrix()
        } else {
            GL11.glPushMatrix()
            GL11.glTranslatef(moveL * 0.07f, 0.0f, doorMoveL * 0.7f)
            model.doorFL.render(renderer)
            GL11.glPopMatrix()

            GL11.glPushMatrix()
            GL11.glTranslatef(moveL * 0.07f, 0.0f, -doorMoveL * 0.7f)
            model.doorBL.render(renderer)
            GL11.glPopMatrix()

            GL11.glPushMatrix()
            GL11.glTranslatef(-moveR * 0.07f, 0.0f, doorMoveR * 0.7f)
            model.doorFR.render(renderer)
            GL11.glPopMatrix()

            GL11.glPushMatrix()
            GL11.glTranslatef(-moveR * 0.07f, 0.0f, -doorMoveR * 0.7f)
            model.doorBR.render(renderer)
            GL11.glPopMatrix()
        }
    }

    @Suppress("DuplicatedCode")
    private fun renderRandomCurtain(renderer: VehiclePartsRenderer) {
        if (renderer.mcTime in 401..12999) return
        ModelART500M.getModel(renderer.modelName)!!.nightWindow.render(renderer)
    }
}
